### ISSUE TITLE
Apache 404 errors are no longer in Apache 2.4.x error_log

### ISSUE THREAD

**Issue by kyz** [fail2ban/fail2ban#2378]
Since Apache 2.4, most people's Apache error_log files no longer have any `File does not exist: filename` lines.

[Apache 2.4 release notes](https://www.apache.org/dist/httpd/CHANGES_2.4): "Downgrade error log messages which accompany a 404 request status from loglevel error to info. [PR 35768](https://bz.apache.org/bugzilla/show_bug.cgi?id=35768)"

Apache can be configured with to show them again with `LogLevel INFO` but that brings in a lot of extra messages to error_log that might be unwanted, and it's also not the default config. By default, the only place these errors will appear now is in `access_log` as 404 errors.

This affects the following filters, all of which look at Apache error log for "does not exist" lines:
* apache-botsearch.conf
* apache-nohome.conf
* apache-noscript.conf

What do you recommend to get these filters matching again, by default?

**Comment by sebres** [comment]
> PR 35768

Hmm... this looks like it is merged (and released) long time ago. How all the years this filters work by all the people? *Nobody uses apache nowadays?* :smile: 

> What do you recommend to get these filters matching again, by default?

Shortly - dunno. 
I would definitely not recommend do a monitoring of the access-log (due to possibly too large "parasite" log-traffic in sense of fail2ban).

I don't use apache (why at all if there is a nginx;), so I don't know how exactly it would be possibly in apache, but I'd recommend to make some customization in order to generate the messages in separate log-file (or journal). For example in nginx it would be something like this - https://github.com/fail2ban/fail2ban/issues/1962#issuecomment-343856328

So we are open for suggestions.

**Comment by 0x6772** [comment]
Todd Vierling (who I think I remember works on Apache? Maybe via Oracle? This guy: https://stackoverflow.com/users/2644135/todd-vierling) outlined pretty clearly how Apache thinks this ought to work these days in the thread starting here: https://www.jwz.org/blog/2019/03/apache-2-4-1-killed-fail2ban-so-thats-awesome/#comment-197663

**Comment by sebres** [comment]
Well as already said I find the way to observe the default logs as not proper (no matter error or access log).
But changing of a default behavior may be also not good practice too.
So I'm partially agree with both sides in the discussion. 

Is there any simple way to configure the log of apache for selected errors like 404/etc per application or common prefix or whatever is called location in apache? And if yes, which logging-format would get this log-messages (access-log I assume).

**Comment by kyz** [comment]
I note that the apache-badbots, apache-fakegooglebot, php-url-fopen and pass2allow-ftp filters already look at the full Apache access_log, so I'm not sure where the performance concern would come from.

As an example, apache-botsearch could change from its current error log searching:

    failregex = ^(?:File does not exist|script not found or unable to stat): <webroot><block>(, referer: \S+)?\s*$
                ^script '<webroot><block>' not found or unable to stat(, referer: \S+)?\s*$

to searching access log instead:

    failregex = /<block>[^ ]* HTTP/\d\.\d" 404 

If this is a performance problem, then the user could add a custom access log that only stores 404 errors, and point fail2ban at that, with this sort of Apache config:

    LogFormat "\"%404r\" %>s " missinglog
    CustomLog logs/404s_only missinglog

This log will either write the URL for a 404 response, or "-" for any other response: 

    "-" 301
    "GET /doesntexist HTTP/1.1" 404
    "-" 400
    "-" 200

**Comment by sebres** [comment]
> so I'm not sure where the performance concern would come from.

The thing is - per default some fail2ban filters are build to monitor a default log-target, what does not say it'd be good practice. Default is even just default. One can and must change it.

Just think about how many entries you'd see on good visited servers (all this legitimate 200th responses etc). So if your access-log shows for each 404 ca. 1000 200th-entries, for fail2ban all this OK-messages are "parasitic" traffic and it produces totally unneeded load on the server.

BTW, see https://github.com/fail2ban/fail2ban/wiki/Best-practice for details.

> This log will either write the URL for a 404 response, or "-" for any other response:

The primary goal was to separate the logs, so 404/etc into "failed.log", all others into "access.log". 
Is it not possible at all in apache?

**Comment by kyz** [comment]
> Just think about how many entries you'd see on good visited servers (all this legitimate 200th responses etc). 

This is true, but there are already fail2ban filters which look at 100% of the access log, e.g. apache-badbots. Hopefully fail2ban is combining filters on the same files, so it only has to read the log once.

>The primary goal was to separate the logs, so 404/etc into "failed.log", all others into "access.log". Is it not possible at all in apache?

Looking into it and testing a bit, it is possible to perform that expression. I've updated my previous comment.

**Comment by yyk123** [comment]
yeah, it logs in access.log now.  and i write my own filter to ban http 4xx error

**Comment by benrubson** [comment]
For performance matters, [conditional logging](https://httpd.apache.org/docs/2.4/en/logs.html) may help you redirection 4xx to a specific logfile.

**Comment by jurgenhaas** [comment]
@kyz what did you end up doing with this one? I've tried your suggestion but couldn't get the RE to work. My log entries look like this:

```
"1.2.3.4" - - [15/Feb/2021:19:07:47 +0100] "GET /MySQLAdmin/index.php HTTP/1.1" 404 229 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/66.0.3359.139 Safari/537.36"
```

The RE `failregex = /<block>[^ ]* HTTP/\d\.\d" 404` doesn't work when I check with

```
fail2ban-regex -o row /var/log/apache2/access.log.1 apache-botsearch
```

The quotes around the IP at the beginning of the line are required in our case as we are behind a reverse proxy and most of the time get a comma separated list of IPs and to keep them together we had to quote them.

**Comment by jurgenhaas** [comment]
Got it to work by removing the `prefregex` line and defining the RE as

```
failregex = ^"<ADDR>".+ HTTP/\d\.\d" 404
```

**Comment by stevepiercy** [comment]
@jurgenhaas @kyz would you please submit a PR with your solutions, even if it is just a comment? It's better than what exists.
